### PR TITLE
DPC-4126: Fix JSON logging and add standardized fields

### PR DESCRIPTION
**Pull request by kyeah**
## 🎫 Ticket

https://jira.cms.gov/browse/DPC-4126

## 🛠 Changes

- Hide Foreman service log prefixes in nonprod (deployed environments)
- Add request-global log attributes for current_user info, organization info, and any raised exceptions
- Fix request_id attribute

## ℹ️ Context for reviewers

Noticed a few issues while auditing the logs produced by the portal in splunk:

- Since we are using foreman to run the fake CPI API GW in nonprod, it is prefixing the service name. This breaks the JSON formatting of each log line, which breaks splunk parsing. Added sed commands to fix this.
- The request_id is not logged because it uses the Rails default `log_tags` functionality. Since we use lograge for request logs, this is not picked up in the logs.
- The normal logs using `Rails.logger` are not printed via JSON.

I also added some helpful information to each request related to the current logged in user, the relevant organization for the API/page request, and any exception details if there was an uncaught exception raised during the request.

## ✅ Acceptance Validation

Verified locally.

Lograge custom log with current_user and organization:

```json
{
  "method": "GET",
  "path": "/portal/organizations/1/invitations/1/accept",
  "format": "html",
  "controller": "InvitationsController",
  "action": "accept",
  "status": 200,
  "allocations": 11634,
  "duration": 72.23,
  "view": 25.18,
  "db": 1.14,
  "ddsource": "ruby",
  "environment": "local",
  "level": "info",
  "request_id": "b225bb9b-72bf-400e-9b28-733fbc9e16e1",
  "request_user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0",
  "request_ip": "REDACTED",
  "current_user": {
    "id": 1,
    "external_id": "60119c5b-606f-4526-b33c-4e691e2d4d3b",
    "pac_id": null
  },
  "organization": {
    "id": 1,
    "dpc_api_organization_id": null,
    "is_authorized_official": false,
    "is_credential_delegate": false
  }
}
```

Custom Rails.logger log:

```json
{
  "level": "INFO",
  "time": "2024-06-19T18:56:02.963+00:00",
  "environment": "local",
  "message": "This is a test",
  "request_id": "16b9c4c6-e84a-4bd0-805f-9a34d966f475",
  "request_user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0",
  "request_ip": "REDACTED",
  "method": "GET",
  "path": "/portal/organizations/1/invitations/1/accept",
  "organization": {
    "id": 1,
    "dpc_api_organization_id": null
  }
}

```

## 🔒 Security Implications

- [ ] This PR adds a new software dependency or dependencies.
- [ ] This PR modifies or invalidates one or more of our security controls.
- [ ] This PR stores or transmits data that was not stored or transmitted before.
- [ ] This PR requires additional review of its security implications for other reasons.

If any security implications apply, add Jason Ashbaugh (GitHub username: StewGoin) as a reviewer and do not merge this PR without his approval.

### CODE REVIEW

File: dpc-portal/config/initializers/logging.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+require './app/lib/dpc_json_logger'
+
+Rails.application.configure do
+    Rails.logger = DpcJsonLogger.new($stdout)
+    config.logger = Rails.logger
+    config.logger.formatter = DpcJsonLogger.formatter
+    config.log_formatter = DpcJsonLogger.formatter
```
Comment (kyeah) [inline dpc-portal/config/initializers/logging.rb:9]: Note that this actually configures it in local dev too, not sure if that's helpful (consistency) or annoying (now you have to read JSON logs.) Happy to move this into the production config file again, but it was helpful for testing and verifying the behavior.

Comment (jdettmannnava) [reply dpc-portal/config/initializers/logging.rb:9]: I would vote no on local dev. The standard logs clump logs related to a request together, and this way makes it harder.

Comment (kyeah) [reply dpc-portal/config/initializers/logging.rb:10]: Updated in df51ba0b. Looks like this now (request logs still in JSON, everything else merged together):

CleanShot 2024-06-21 at 11 37 28@2x

File: dpc-portal/spec/lib/dpc_json_logger_spec.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+RSpec.describe DpcJsonLogger do
+  let(:strout) { StringIO.new }
+  let(:logger) do
+    lgr = DpcJsonLogger.new(strout)
+    lgr.formatter = DpcJsonLogger.formatter
+    lgr
+  end
+
+  describe :info do
```
Comment (jdettmannnava) [inline dpc-portal/spec/lib/dpc_json_logger_spec.rb:13]: I was just playing with RSpec.shared_examples, and I think they would be good here. You wouldn't need a separate file for them.

File: dpc-portal/app/models/current_attributes.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    CurrentAttributes.organization = {
+      id: org.id,
+      dpc_api_organization_id: org.dpc_api_organization_id
+    }
+
+    return unless user
+
+    begin
+      CurrentAttributes.organization[:is_authorized_official] = user.ao?(org)
```
Comment (jdettmannnava) [inline dpc-portal/app/models/current_attributes.rb:41]: I think maybe a role attribute would be better? ao|cd|nil

File: dpc-portal/spec/lib/dpc_json_logger_spec.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe DpcJsonLogger do
+  let(:strout) { StringIO.new }
```
Comment (jdettmannnava) [inline dpc-portal/spec/lib/dpc_json_logger_spec.rb:6]: stdout?

Comment (kyeah) [reply dpc-portal/spec/lib/dpc_json_logger_spec.rb:6]: it's a string output, so `strout` 🤷 🙃

File: dpc-portal/spec/lib/dpc_json_logger_spec.rb (hunk 2 above)
Comment (kyeah) [reply dpc-portal/spec/lib/dpc_json_logger_spec.rb]: ah nice, worked on the first try 😄 switched in https://github.com/CMSgov/dpc-app/pull/2198/commits/f4fc7c1f78e8f4107d1470f241989baf2bf2296d

File: dpc-portal/app/models/current_attributes.rb (hunk 3 above)
Comment (kyeah) [reply dpc-portal/app/models/current_attributes.rb]: Switched to `current_user_role` in https://github.com/CMSgov/dpc-app/pull/2198/commits/0fe3f9469bb8a55b4f4033b2658828929f0c52f5 👍

File: dpc-portal/app/lib/dpc_json_logger.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  def warn(*msg, &)
+    value = as_hash(msg[0], msg[1], &)
+    super(value, &nil)
+  end
+
+  def error(*msg, &)
+    value = as_hash(msg[0], msg[1], &)
+    super(value, &nil)
+  end
+
```
Comment (jdettmannnava) [inline dpc-portal/app/lib/dpc_json_logger.rb:38]: Sorry, just realized missing :fatal and :unknown

**Pull request merged**